pkg/validation/variables/variables_command_test.go

```diff
@@ -0,0 +1,149 @@
+				t.Errorf("Got unexpected error: %s", err)
+			} else {
+				expectedExecCommand := v1alpha2.ExecCommand{}
+				readFileToStruct(t, tt.outputFile, &expectedExecCommand)
+				assert.Equal(t, expectedExecCommand, testExecCommand, "The two values should be the same.")
+			}
+		})
+	}
+}
+
+func TestValidateAndReplaceCompositeCommand(t *testing.T) {
+
+	tests := []struct {
+		name         string
+		testFile     string
+		outputFile   string
+		variableFile string
+		wantErr      bool
+	}{
+		{
+			name:         "Good Substitution",
+			testFile:     "test-fixtures/commands/composite.yaml",
+			outputFile:   "test-fixtures/commands/composite-output.yaml",
+			variableFile: "test-fixtures/variables/variables-referenced.yaml",
+			wantErr:      false,
+		},
+		{
+			name:         "Invalid Reference",
+			testFile:     "test-fixtures/commands/composite.yaml",
+			outputFile:   "test-fixtures/commands/composite.yaml",
+			variableFile: "test-fixtures/variables/variables-notreferenced.yaml",
+			wantErr:      true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			testCompositeCommand := v1alpha2.CompositeCommand{}
+			readFileToStruct(t, tt.testFile, &testCompositeCommand)
+
+			testVariable := make(map[string]string)
+			readFileToStruct(t, tt.variableFile, &testVariable)
+
+			err := validateAndReplaceForCompositeCommand(testVariable, &testCompositeCommand)
+			verr, ok := err.(*InvalidKeysError)
+			if tt.wantErr && !ok {
+				t.Errorf("Expected InvalidKeysError error from test but got %+v", verr)
+			} else if !tt.wantErr && err != nil {
+				t.Errorf("Got unexpected error: %s", err)
+			} else {
+				expectedCompositeCommand := v1alpha2.CompositeCommand{}
+				readFileToStruct(t, tt.outputFile, &expectedCompositeCommand)
+				assert.Equal(t, expectedCompositeCommand, testCompositeCommand, "The two values should be the same.")
+			}
+		})
+	}
+}
+
+func TestValidateAndReplaceApplyCommand(t *testing.T) {
+
+	tests := []struct {
+		name         string
+		testFile     string
+		outputFile   string
+		variableFile string
+		wantErr      bool
+	}{
+		{
+			name:         "Good Substitution",
+			testFile:     "test-fixtures/commands/apply.yaml",
+			outputFile:   "test-fixtures/commands/apply-output.yaml",
+			variableFile: "test-fixtures/variables/variables-referenced.yaml",
+			wantErr:      false,
+		},
+		{
+			name:         "Invalid Reference",
+			testFile:     "test-fixtures/commands/apply.yaml",
+			outputFile:   "test-fixtures/commands/apply.yaml",
+			variableFile: "test-fixtures/variables/variables-notreferenced.yaml",
+			wantErr:      true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			testApplyCommand := v1alpha2.ApplyCommand{}
+			readFileToStruct(t, tt.testFile, &testApplyCommand)
+
+			testVariable := make(map[string]string)
+			readFileToStruct(t, tt.variableFile, &testVariable)
+
+			err := validateAndReplaceForApplyCommand(testVariable, &testApplyCommand)
```

Comment:
I don't see where we actually check the value of `err`. The test requires `outputFile` to be identical to `testFile` when we expect an error, which seems like a strange choice.

Reply:
I dont necessarily check the contents of the err. I just make sure its `verr, ok := err.(*InvalidKeysError)` in the next line. As for the output file, since we decided that any unresolved substitution should remain as it is, I just make sure the devfile is as expected that is they have `{{invalid-var}}` when its unresolved.

Reply:
FWIW I have a test that does check the warning struct where the keys come from the errors. The test is in https://github.com/devfile/api/pull/352/files#diff-88492713a9572b112b36ac53c64c195ed31f959b4835a58d7e985d96cfecd912R31-R50

Reply:
:+1: makes sense to me.